node/core/candidate-validation/src/lib.rs

```diff
@@ -398,16 +417,51 @@ async fn validate_candidate_exhaustive(
 		relay_parent_storage_root: persisted_validation_data.relay_parent_storage_root,
 	};
 
-	let result = validation_backend
-		.validate_candidate(raw_validation_code.to_vec(), params)
-		.await;
+	let mut result = validation_backend.validate_candidate(validation_code, params.clone()).await;
 
 	if let Err(ref e) = result {
 		tracing::debug!(
 			target: LOG_TARGET,
 			error = ?e,
 			"Failed to validate candidate",
 		);
+
+		// In case preimage for the supplied code hash was not found by the
+		// validation host, request the code from Runtime API and try again.
+		if let &ValidationError::ArtifactNotFound = e {
+			tracing::debug!(
+				target: LOG_TARGET,
+				"Validation host failed to find artifact by provided hash",
+			);
+
+			let validation_code = match request_validation_code_by_hash(ctx, &descriptor).await? {
+				Ok(Some(validation_code)) => validation_code,
+				Ok(None) =>
+					return Ok(Err(ValidationFailed(
+						"Runtime API didn't return validation code by hash".into(),
+					))),
+				Err(err) => return Ok(Err(ValidationFailed(err.to_string()))),
+			};
+
+			let validation_code = Pvf::from_code(
+				match sp_maybe_compressed_blob::decompress(
+					&validation_code.0,
+					VALIDATION_CODE_BOMB_LIMIT,
+				) {
+					Ok(code) => code,
+					Err(e) => {
+						tracing::debug!(target: LOG_TARGET, err=?e, "Invalid validation code");
+
+						// If the validation code is invalid, the candidate certainly is.
+						return Ok(Ok(ValidationResult::Invalid(
+							InvalidCandidate::CodeDecompressionFailure,
+						)))
+					},
+				}
+				.to_vec(),
+			);
```

Comment:
Nit: it's getting visually hard to get a mental structure of this code when reading. First there is a call to `decompress`, then either break the control flow or yield `code`. If we returned code we `to_vec` it and then do `from_code`.

I think if we extracted the actual argument to `from_code` into a separate variable and moved `to_vec` closer. The end result would be a more straightforward code like, ideally just top-to-bottom.